exercises/matthew_cumber/ex01_basics/src/main.cpp

```diff
@@ -0,0 +1,199 @@
+      w.incrementCount();
+      return true;
+    }
+  }
+
+  return false;
+}
+
+/** Tests if a letter is one of the predefined punctuation characters
+ * @param letter The letter which is being tested
+ * @return A boolean, true if a letter is a punctuation character, false otherwise
+ */
+bool isPunctuation(const char letter)
+{
+  for(int i = 0; i < PUNCTUATION_COUNT; ++i)
+  {
+    if(letter == PUNCTUATION[i]) return true;
+  }
+
+  return false;
+}
+
+/** Writes the unique words to an output file "UniqueWords.txt"
+ * @param uniqueWords The vector of unique words to be written to the file
+ * @return no return value
+ */
+void writeUniqueWordsToFile(std::vector<UniqueWord> const uniqueWords)
+{
+  // Try to opne the output file
+  std::ofstream file(PATH+"/src/output.txt");
+  if(!file.is_open())
+  {
+    std::cout << "Error opening file \"output.txt\", exiting ..." << std::endl;
+    return;
+  }
+
+  // Header of output file
+  file << "Word" << std::setw(FIELD_WIDTH) << "Usage" << std::endl << std::endl;
+
+  int fieldWidth = 10;
+
+  // To write words in descending order need to loop vector in reverse
+  for(auto w = uniqueWords.rbegin(); w != uniqueWords.rend(); ++w)
+  {
+    // Format the output to align count values
+    fieldWidth = FIELD_WIDTH-(int)w->getWord().length();
+    if(fieldWidth <= 0) file << w->getWord() << " " << w->getCount() << std::endl;
+    else file << w->getWord() << std::setw(fieldWidth) << w->getCount() << std::endl;
+  }
+
+  file.close();
+
+  std::cout << "Words successfuly saved in \"output.txt\"" << std::endl;
+
+  return;
+}
+
+int main(int argc, char const **argv)
+{
+  std::vector<UniqueWord> uniqueWords; // Stores words and counts as objects in a vector
+
+	// Check a file path has been provided as an argument on cl
+	if(argc < 2)
+	{
+		std::cout << "Please provide a path to the file to be read as a command line argument." << std::endl;
+    std::cout << "Example usage:" << std::endl;
+    std::cout << "  ./WordCounter \"example.txt\"" << std::endl;
+    return 1;
+	}
+
+  // Try to open file from current directory, then try as if path passed as argument
+  std::ifstream file(PATH+"/tst_files/"+argv[1]);
+  if(!file.is_open())
+  {
+    file.open(argv[1]);
+    if(!file.is_open())
+    {
+      std::cout << "Error opening file \"" + std::string(argv[1]) + "\", exiting ..." << std::endl;
+      return 1;
+    }
+  }
+  
+  std::cout << "Reading file " + std::string(argv[1]) << std::endl;
+
+  std::string word;
+
+  // Loop file word by word, reads characters spearated by spaces
+  while(file >> word)
+  {
+    // Skip words less than five characters
+    if(word.length() < 5) continue;
+
+    // Remove Punctuation, Convert all letters to lower case, and Remove hyphens
+    for(int i = 0; i < word.length(); ++i)
+    {
+      // Erase letter if punctuation
+      while(isPunctuation(word[i]))
+      {
+        word.erase(i,1);
+        if(i >= word.length()) break; // Check not at end of word
```

Comment:
This comment is a bit subjective - but it feels like there's a bit too much attempted in this loop through the characters of "word". I think it's functionally OK but it looks quite complicated and could be difficult to maintain.

Please can you have a go at rearranging this a bit so that you don't need this kind of "check not at end of word" line. Also would be nice if the logic to check the word length is >4 and then add the word into the list could be done in one place - currently it's repeated at L149 and L166

Perhaps you could manage the hyphens in their own loop rather than trying to do it in the same pass as removing punctuation\converting to lower case. You can split the word variable returned from "file>>word" into further subwords using the std::getline function. Something like:

        std::istringstream ss(word);
        std::string wordWithoutHyphens;
        while (std::getline(ss, wordWithoutHyphens, '-')) {
        }